### PR TITLE
Hparams: use data_provider.read_last_scalars directly

### PR DESCRIPTION
**Pull request by yatbear**
Use `data_provider.read_last_scalars` directly instead of using `read_scalars` with `downsample` set to 1.

#hparams

### CODE REVIEW

**Comment by yatbear** [comment]
cc @bmd3k

File: tensorboard/plugins/hparams/backend_context.py [hunk 1]
```diff
@@ -178,26 +178,19 @@ def read_last_scalars(self, ctx, experiment_id, run_tag_filter):
...
-            the semantics as in `read_scalars`.
+            the semantics as in `read_last_scalars`.
 
         Returns:
           A dict `d` such that `d[run][tag]` is a `provider.ScalarDatum`
           value, with keys only for runs and tags that actually had
           data, which may be a subset of what was requested.
         """
-        data_provider_output = self._tb_context.data_provider.read_scalars(
+        return self._tb_context.data_provider.read_last_scalars(
```
Comment (arcra) [inline tensorboard/plugins/hparams/backend_context.py:188]: Should we just remove this method, and call the DP directly from wherever this is called? Or did you explicitly prefer to keep this method? Just checking if you considered that, I'm ok with this too.

**Review by arcra — APPROVED** [review]

File: tensorboard/plugins/hparams/backend_context.py (hunk 1 above)
Comment (yatbear) [reply tensorboard/plugins/hparams/backend_context.py]: Great idea! Done.

Comment (arcra) [reply tensorboard/plugins/hparams/backend_context.py]: I'm just realizing there are similar wrapping methods here that just call the DP internally, such as hparams_from_data_provider; although they explicitly mention data_provider in the name, and I believe only the new list/read hparams methods are the ones that are simply wrapped in a forwarding call.

It might be a good idea to make it consistent by either keeping the wrapping method that we had, or by also removing the other wrapping methods, and updating the calls to them to call the DP's methods. WDYT?

I'm not sure if the wrapping methods exist because the naming and semantics are more appropriate than the DP methods (?), or perhaps it was mostly because the DP was not a publicly accessible member before, so wrapping it was an easy way to just add that functionality without worrying/deciding to expose that as a public member. It might be worth checking if Brian has an opinion here.

Sorry for the churn.

Comment (yatbear) [reply tensorboard/plugins/hparams/backend_context.py]: NW! cc'ing @bmd3k to see if there's any preference.

Comment (yatbear) [reply tensorboard/plugins/hparams/backend_context.py]: @arcra I just realized the last comment was left on an outdated file version. I replaced all the `self._tb_context.data_provider` with `self._data_provider` in the latest commit, PTAAL and LMK if it looks okay, thanks!

Comment (arcra) [reply tensorboard/plugins/hparams/backend_context.py]: Not sure if I had actually seen a different snapshot or not, but the general idea from my last comment still applies.

I think the "backend_context" still has a couple methods that just wrap/forward a single call to the DataProvider's methods (namely, list/read hparams methods), and I'm not sure if the intention was to *not* expose the DataProvider as a member of the backend_context object.

Even if that wasn't necessarily the intention, I think it would be a bit better if we consistently either have wrapping methods for calling the DP's methods, or remove any wrapping methods that are only forwarding that call to the DP's methods (there are a couple other "wrapping" methods that are doing some additional work on the response from the DP call, but I'm referring to the ones that simply return the result directly).

I lean towards adding back the wrapping method that we had, and avoid exposing the DP instance through that `@property` that you added. Even tho the DP is supposed to be a "public" interface for plugins to use, exposing it as a member of the `backend_context` object feels like it's leaking from the encapsulation from that object.

In short, sorry for the churn, what you had originally with a wrapping method that calls the DP makes sense to me. (But overall, either way is probably fine, won't make a big difference)

Comment (yatbear) [reply tensorboard/plugins/hparams/backend_context.py:188]: NW! Thanks for the review! I reverted the last commit. I also think it's safer not to expose data_provider in case it causes any unexpected breakage internally later.

Comment (arcra) [reply tensorboard/plugins/hparams/backend_context.py:188]: Thanks! LGTM.
